### PR TITLE
MEPTS 1294 | TB7 Advanced Disease TB Cascade

### PR DESCRIPTION
**Pull request by eliazmuchave**
The DAH enconter metadata (encouter type 90) should be referenced in the DAH report that has not been reviewed yet

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java [hunk 1]
```diff
@@ -0,0 +1,1485 @@
...
+            + "                                            AND o.voided = 0 "
+            + "                                            AND e.encounter_type = ${6} "
+            + "                                            AND o.concept_id = ${6273} "
+            + "                                            AND e.location_id = :location "
+            + "                                            AND e.encounter_datetime BETWEEN :startDate AND :endDate "
+            + "                                     GROUP  BY e.patient_id) recent_state "
+            + "                                  ON recent_state.patient_id = e.patient_id "
+            + "                   WHERE  e.voided = 0 "
+            + "                          AND o.voided = 0 "
+            + "                          AND o.concept_id = ${6273} "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java:601]: @eliazmuchave  I think here is missing the encounter_type

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java [hunk 2]
```diff
@@ -0,0 +1,1485 @@
...
+            + "                                     WHERE  e.voided = 0 "
+            + "                                            AND o.voided = 0 "
+            + "                                            AND e.encounter_type = ${53} "
+            + "                                            AND o.concept_id = ${6272} "
+            + "                                            AND e.location_id = :location "
+            + "                                            AND o.obs_datetime BETWEEN :startDate AND :endDate "
+            + "                                     GROUP  BY e.patient_id) recent_state ON recent_state.patient_id = e.patient_id "
+            + "                   WHERE  e.voided = 0 "
+            + "                          AND o.voided = 0 "
+            + "                          AND o.concept_id = ${6272} "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java:624]: @eliazmuchave same here, encounter_type

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java [hunk 3]
```diff
@@ -0,0 +1,1485 @@
...
+    SqlCohortDefinition cd = new SqlCohortDefinition();
+    cd.setName("Number of clients with a CD4 count during inclusion period");
+    cd.addParameter(new Parameter("startDate", "Start Date", Date.class));
+    cd.addParameter(new Parameter("endDate", "End Date", Date.class));
+    cd.addParameter(new Parameter("location", "End Date", Location.class));
+
+    String query =
+        "SELECT p.patient_id "
+            + "FROM   patient p "
+            + "       INNER JOIN (SELECT e.patient_id, Max(DATE(e.encounter_datetime)) AS result_date "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java:670]: Acho que não precisas deste maximo

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java [hunk 4]
```diff
@@ -0,0 +1,1485 @@
...
+            + "                   WHERE  e.voided = 0 "
+            + "                          AND o.voided = 0 "
+            + "                          AND e.encounter_type IN ( ${6}, ${13}, ${51} ) "
+            + "                          AND o.concept_id = ${1695} "
+            + "                          AND o.value_numeric IS NOT NULL "
+            + "                          AND e.location_id = :location "
+            + "                          AND DATE(e.encounter_datetime) BETWEEN :startDate AND :endDate "
+            + "                   GROUP  BY e.patient_id "
+            + "                   UNION "
+            + "                   SELECT e.patient_id, Max(o.obs_datetime) AS result_date "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java:682]: Aqui tb

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java [hunk 5]
```diff
@@ -0,0 +1,1485 @@
...
+    cd.addParameter(new Parameter("endDate", "End Date", Date.class));
+    cd.addParameter(new Parameter("location", "End Date", Location.class));
+
+    String query =
+        "SELECT p.patient_id "
+            + "FROM   patient p "
+            + "       INNER JOIN encounter e ON p.patient_id = e.patient_id "
+            + "       INNER JOIN obs o ON o.encounter_id = e.encounter_id "
+            + "       INNER JOIN (SELECT recent_cd4.patient_id, Max(cd4_date) recent_date "
+            + "                   FROM  (SELECT e.patient_id, Date(e.encounter_datetime) cd4_date "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java:1280]: Sera que aqui neste union não estas a precisar de 2 maxs para depois tirar o maximo deles? Não tem spec

**Review by psmatsinhe — CHANGES_REQUESTED** [review]
@eliazmuchave please check my comments

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java (hunk 1 above)
Comment (eliazmuchave) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java:601]: Fixed

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java (hunk 2 above)
Comment (eliazmuchave) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java:624]: Fixed

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java (hunk 3 above)
Comment (eliazmuchave) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java:670]: Will be useful for debugging, as the specification refers to the most recent date.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java (hunk 4 above)
Comment (eliazmuchave) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java:682]: Will be useful for debugging, as the specification refers to the most recent date

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java (hunk 5 above)
Comment (eliazmuchave) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java:1280]: O máximo é do resultado do UNION é tirado na linha 1277

**Review by psmatsinhe — APPROVED** [review]
Looks good to me

**Pull request merged**